frame/offences/src/migration.rs

```diff
@@ -0,0 +1,122 @@
+    );
+
+    /// Deferred reports that have been rejected by the offence handler and need to be submitted
+    /// at a later time.
+    pub struct DeferredOffences<T: Config>(PhantomData<T>);
+
+    impl<T: Config> StorageValue<Vec<DeferredOffenceOf<T>>> for DeferredOffences<T> {
+        type Query = Vec<DeferredOffenceOf<T>>;
+        fn module_prefix() -> &'static [u8] {
+            b"Offences"
+        }
+        fn storage_prefix() -> &'static [u8] {
+            b"DeferredOffences"
+        }
+        fn from_optional_value_to_query(v: Option<Vec<DeferredOffenceOf<T>>>) -> Self::Query {
+            v.unwrap_or_else(|| Default::default())
+        }
+        fn from_query_to_optional_value(v: Self::Query) -> Option<Vec<DeferredOffenceOf<T>>> {
+            Some(v)
+        }
+    }
+}
+
+pub fn remove_deferred_storage<T: Config>() -> Weight {
+    let mut weight = T::DbWeight::get().reads_writes(1, 1);
+    let deferred = <deprecated::DeferredOffences<T>>::take();
+    log::info!(target: "runtime::offences", "have {} deferred offences, applying.", deferred.len());
+    for (offences, perbill, session) in deferred.iter() {
+        let consumed = T::OnOffenceHandler::on_offence(&offences, &perbill, *session);
+        weight = weight.saturating_add(consumed);
+    }
+
+    weight
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+    use crate::mock::{new_test_ext, with_on_offence_fractions, Offences, Runtime as T, System};
+    use frame_support::traits::OnRuntimeUpgrade;
+    use sp_runtime::Perbill;
+    use sp_staking::offence::OffenceDetails;
+
+    #[test]
+    fn should_resubmit_deferred_offences() {
+        new_test_ext().execute_with(|| {
+            // given
+            assert_eq!(<deprecated::DeferredOffences<T>>::get().len(), 0);
+            with_on_offence_fractions(|f| {
+                assert_eq!(f.clone(), vec![]);
+            });
+
+            let offence_details = OffenceDetails::<
+                <T as frame_system::Config>::AccountId,
+                <T as Config>::IdentificationTuple,
+            > {
+                offender: 5,
+                reporters: vec![],
+            };
+
+            // push deferred offence
+            <deprecated::DeferredOffences<T>>::mutate(|d| {
+                d.push((
+                    vec![offence_details],
+                    vec![Perbill::from_percent(5 + 1 * 100 / 5)],
+                    1,
+                ))
+            });
+
+            // when
+            assert_eq!(
+                Offences::on_runtime_upgrade(),
+                <T as frame_system::Config>::DbWeight::get().reads_writes(1, 2),
+            );
+
+            // then
+            assert_eq!(<deprecated::DeferredOffences<T>>::get().len(), 0);
+            with_on_offence_fractions(|f| {
+                assert_eq!(f.clone(), vec![Perbill::from_percent(5 + 1 * 100 / 5)]);
+            });
+
+            // No events emitted
+            assert_eq!(System::events().len(), 0);
```

Comment:
```suggestion
```

I see the thinking, but a bit too superfluous. Checking that the queued stuff got processed is enough.